### PR TITLE
Response with instance

### PR DESCRIPTION
**Pull request by sradack**
Note: this is based on #232

## Changes proposed in this PR

- Refactors process, includes `:instance` in the response, and changes the exception handling

## Why are we making these changes?

- The instance in the response allows more debug headers to be centralized in `wrap-debug`.
- `process` takes fewer parameters, has fewer responsibilities

### CODE REVIEW

File: waiter/src/waiter/process_request.clj [hunk 1]
```diff
@@ -584,37 +589,32 @@
...
                         (catch Exception e
                           ; A :client-error or :instance-error may already be in the channel in which case our :generic-error will be ignored.
                           (deliver reservation-status-promise :generic-error)
                           ; close request-state-chan to mark the request as finished
                           (async/close! request-state-chan)
-                          ; Handle error lower down
-                          (throw e))))))
+                          (-> (handle-process-exception e request @response-headers)
+                              (assoc :instance instance)))))))
                 (catch Exception e
```
[the remark this answers is not in the document]
Comment (sradack) [reply waiter/src/waiter/process_request.clj]: `prepare-instance` can still throw.

[the remark this answers is not in the document]
Comment (sradack) [reply waiter/src/waiter/process_request.clj]: I reworked the exception handling a bit, but kept the try block.

File: waiter/src/waiter/process_request.clj [hunk 2]
```diff
@@ -567,8 +574,6 @@
                                                          add-debug-header-into-response!))]
                       (try
                         (log/info "suggested instance:" (:id instance) (:host instance) (:port instance))
-                        (when waiter-debug-enabled?
-                          (swap! response-headers merge (instance->debug-headers instance prepend-waiter-url)))
```
[the remark this answers is not in the document]
Comment (sradack) [reply waiter/src/waiter/process_request.clj]: Removing

File: waiter/src/waiter/websocket.clj [hunk 3]
```diff
@@ -354,16 +355,16 @@
                   (log/debug "ignoring exception generated from closed connection")))))))
       (catch Exception e
         (async/>!! request-close-promise-chan :process-error)
-        (throw e)))))
+        (throw e))))
```
[the remark this answers is not in the document]
Comment (sradack) [reply waiter/src/waiter/websocket.clj]: Changed to log statement

File: waiter/src/waiter/process_request.clj [hunk 4]
```diff
@@ -584,37 +589,32 @@
                               confirm-live-connection-with-abort (confirm-live-connection-factory request-abort-callback)]
                           (when error
                             (throw-response-error error reservation-status-promise instance response-headers))
-                          (process-backend-response-fn local-usage-agent instance-request-properties descriptor instance request
-                                                       reason-map response-headers reservation-status-promise
-                                                       confirm-live-connection-with-abort request-state-chan response))
+                          (-> (process-backend-response-fn local-usage-agent instance-request-properties descriptor instance request
+                                                           reason-map response-headers reservation-status-promise
+                                                           confirm-live-connection-with-abort request-state-chan response)
+                              (assoc :instance instance)))
```
[the remark this answers is not in the document]
Comment (sradack) [reply waiter/src/waiter/process_request.clj]: Lovely

File: waiter/src/waiter/core.clj [hunk 5]
```diff
@@ -177,18 +177,29 @@
...
-                          (update response :headers
-                                  (fn [headers]
-                                    (-> headers
-                                        (assoc "x-waiter-request-date" (utils/date-to-str request-time utils/formatter-rfc822))
-                                        (assoc "x-waiter-request-id" request-id)
-                                        (assoc "x-waiter-router-id" router-id)))))]
+            add-headers (fn [{:keys [descriptor instance] :as response}]
+                          (let [backend-directory (:log-directory instance)
+                                backend-log-url (and backend-directory
+                                                     (handler/generate-log-url prepend-waiter-url instance))
```
[the remark this answers is not in the document]
Comment (sradack) [reply waiter/src/waiter/core.clj]: Done

File: waiter/src/waiter/core.clj [hunk 6]
```diff
@@ -177,18 +177,29 @@
...
+                                request-date (and request-time (utils/date-to-str request-time utils/formatter-rfc822))]
```
[the remark this answers is not in the document]
Comment (sradack) [reply waiter/src/waiter/core.clj]: Haha whoops, yes of course.  My bad.  `and` did what I wanted though.

File: waiter/src/waiter/core.clj [hunk 7]
```diff
@@ -177,18 +177,29 @@
...
-            add-headers (fn [response]
-                          (update response :headers
-                                  (fn [headers]
-                                    (-> headers
-                                        (assoc "x-waiter-request-date" (utils/date-to-str request-time utils/formatter-rfc822))
-                                        (assoc "x-waiter-request-id" request-id)
-                                        (assoc "x-waiter-router-id" router-id)))))]
+            add-headers (fn [{:keys [descriptor instance] :as response}]
+                          (let [backend-directory (:log-directory instance)
+                                backend-log-url (and backend-directory
```
[the remark this answers is not in the document]
Comment (sradack) [reply waiter/src/waiter/core.clj]: Done

File: waiter/src/waiter/process_request.clj [hunk 8]
```diff
@@ -533,81 +517,68 @@
...
-                (catch Exception e
-                  ; Handle error lower down
-                  (throw (ex-info "Rethrowing exception from process" {:descriptor descriptor :source :pr/process} e)))))
-            (catch Exception e
-              (let [{:keys [descriptor source]} (ex-data e)
-                    exception (if (= source :pr/process) (.getCause e) e)]
-                (-> (process-exception-fn track-process-error-metrics request exception)
-                    (update :headers (fn [headers]
-                                       (merge @response-headers headers))))))))))))
+          (let [{:keys [service-id service-description]} descriptor
```
Comment (sradack) [inline waiter/src/waiter/process_request.clj]: Removed a call to `confirm-live-connection-without-abort` and the surrounding `try`/`catch` block.

File: waiter/src/waiter/process_request.clj [hunk 9]
```diff
@@ -533,81 +517,68 @@
...
+                                            :request-id request-id}
+                                     priority (assoc :priority priority))
+                        ; pass false to keep request-state-chan open after control-mult is closed
+                        ; request-state-chan should be explicitly closed after the request finishes processing
+                        request-state-chan (async/tap control-mult (au/latest-chan) false)
+                        queue-timeout-ms (:queue-timeout-ms instance-request-properties)
+                        instance (<? (prepare-instance instance-rpc-chan service-id reason-map start-new-service-fn request-state-chan
+                                                       queue-timeout-ms reservation-status-promise metric-group
+                                                       add-debug-header-into-response!))]
+                    (-> (try
```
Comment (sradack) [inline waiter/src/waiter/process_request.clj]: Only one assoc for `:instance`

File: waiter/src/waiter/process_request.clj [hunk 10]
```diff
@@ -533,81 +517,68 @@
...
+                                                         confirm-live-connection-with-abort request-state-chan response))
+                          (catch Exception e
+                            ; A :client-error or :instance-error may already be in the channel in which case our :generic-error will be ignored.
+                            (deliver reservation-status-promise :generic-error)
+                            ; close request-state-chan to mark the request as finished
+                            (async/close! request-state-chan)
+                            (handle-process-exception e request @response-headers)))
+                        (assoc :instance instance)))))
+              (catch Exception e ; Handle case where we couldn't get an instance
+                (counters/dec! (metrics/service-counter service-id "request-counts" "outstanding"))
```
Comment (sradack) [inline waiter/src/waiter/process_request.clj]: Metrics updates taken out of `ex-handler` and moved here.

File: waiter/src/waiter/websocket.clj [hunk 11]
```diff
@@ -354,16 +355,16 @@
                   (log/debug "ignoring exception generated from closed connection")))))))
       (catch Exception e
         (async/>!! request-close-promise-chan :process-error)
-        (throw e)))))
+        (log/error e "error while processing websocket response"))))
```
Comment (sradack) [inline waiter/src/waiter/websocket.clj]: No longer throwing exception

File: waiter/src/waiter/process_request.clj [hunk 12]
```diff
@@ -134,46 +134,41 @@
...
-                                (cid/cerror correlation-id "reservation status not received!"))
-                              (when (not= :success-async status)
-                                (counters/dec! (metrics/service-counter service-id "request-counts" "outstanding"))
-                                (statsd/gauge-delta! metric-group "request_outstanding" -1))
-                              (service/release-instance-go instance-rpc-chan instance {:status status, :cid correlation-id, :request-id request-id})))
-                          (fn [e] (log/error e "error releasing instance!")))
-        instance)
-      (catch Exception e
-        (ex-handler e)))))
+  (go-try
```
Comment (sradack) [inline waiter/src/waiter/process_request.clj]: Removed `ex-handler`.  Combined `(go (try ..))` into `(go-try ..)`  Code for `ex-handler` moved to `catch` block that calls `prepare-instance`

**Comment by shamsimam** [comment]
Please rebase.

**Comment by sradack** [comment]
Rebased

File: waiter/src/waiter/process_request.clj [hunk 13]
```diff
@@ -533,81 +516,68 @@
...
-                  (throw (ex-info "Rethrowing exception from process" {:descriptor descriptor :source :pr/process} e)))))
-            (catch Exception e
-              (let [{:keys [descriptor source]} (ex-data e)
-                    exception (if (= source :pr/process) (.getCause e) e)]
-                (-> (process-exception-fn track-process-error-metrics request exception)
-                    (update :headers (fn [headers]
-                                       (merge @response-headers headers))))))))))))
+          (let [{:keys [service-id service-description]} descriptor
+                {:strs [metric-group]} service-description]
+            (send local-usage-agent metrics/update-last-request-time-usage-metric service-id request-time)
```
Comment (shamsimam) [inline waiter/src/waiter/process_request.clj]: For a different PR: this can be moved to another middleware.

**Pull request merged**